### ISSUE TITLE
different behaviors of sonnet module

### ISSUE THREAD

**Issue by mingyr** [deepmind/sonnet#84]
Dear Deepminder:

Actually it is a complicated issue, please stay patient.

I want to wrap the tf.metrics around a Sonnet module for measuring performance of each batch, and the following is the work I have done:

```
class Metrics(snt.AbstractModule):
    def __init__(self, indicator, summaries = None, name = "metrics"):
        super(Metrics, self).__init__(name = name)
        self._indicator = indicator
        self._summaries = summaries

    def _build(self, labels, logits):
        if self._indicator == "accuracy":
            metric, metric_update = tf.metrics.accuracy(labels, logits)
            with tf.control_dependencies([metric_update]):
                outputs = tf.identity(metric)
        elif self._indicator == "precision":
            metric, metric_update = tf.metrics.precision(labels, logits)
            with tf.control_dependencies([metric_update]):
                outputs = tf.identity(metric)
        elif self._indicator == "recall":
            metric, metric_update = tf.metrics.recall(labels, logits)
            with tf.control_dependencies([metric_update]):
                outputs = tf.identity(metric)
        elif self._indicator == "f1_score":
            metric_recall, metric_update_recall = tf.metrics.recall(labels, logits)
            metric_precision, metric_update_precision = tf.metrics.precision(labels, logits)
            with tf.control_dependencies([metric_update_recall, metric_update_precision]):
                outputs = 2.0 / (1.0 / metric_recall + 1.0 / metric_precision)
        else:
            raise ValueError("unsupported metrics")

        if type(self._summaries) == list:
            self._summaries.append(tf.summary.scalar(self._indicator, outputs))

        return outputs

```

Since I have to reset the internal state for each batch, so I also write some utility:

```
def reset_metrics(sess, debug = False):
    lvars = tf.get_collection(tf.GraphKeys.LOCAL_VARIABLES, scope = "metrics*")
    lvars_initializer = tf.variables_initializer(var_list = lvars)
    sess.run(lvars_initializer)

    if debug:
        variables_names = [v.name for v in lvars]
        values = sess.run(variables_names)
        for k, v in zip(variables_names, values):
            print "Variable: ", k
            print "Shape: ", v.shape
            print v

```

However, when I want to test the module, the following code  works:
```
def test3():
    import numpy as np

    labels = tf.constant([1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], tf.int32)
    logits = tf.constant([1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], tf.int32)

    metrics = Metrics("accuracy")
    accuracy = metrics(labels, logits)

    metrics2 = Metrics("f1_score")
    f1_score = metrics2(labels, logits)

    writer = tf.summary.FileWriter("utils-const", tf.get_default_graph())
    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer(), tf.local_variables_initializer()])

        accu, f1 = sess.run([accuracy, f1_score])
        print(accu)
        print(f1)

    writer.close()

```
However the following code does NOT work

```
def test4():
    from tensorflow.python import debug as tf_debug
    import numpy as np

    tf_labels = tf.placeholder(dtype=tf.int32, shape=[None])
    tf_logits = tf.placeholder(dtype=tf.int32, shape=[None])

    labels = np.array([1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], np.int32)
    logits = np.array([1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], np.int32)

    metrics = Metrics("accuracy")
    accuracy = metrics(tf_labels, tf_logits)

    metrics2 = Metrics("f1_score")
    f1_score = metrics2(tf_labels, tf_logits)

    writer = tf.summary.FileWriter("utils-feed", tf.get_default_graph())
    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer(), tf.local_variables_initializer()])

        sess = tf_debug.LocalCLIDebugWrapperSession(sess)

        accu, f1 = sess.run([accuracy, f1_score], feed_dict = {tf_labels: labels, tf_logits: logits})
        print(accu)
        print(f1)

    writer.close()

```

However, the routine test3() is equivalent to test4(), isn't it?

So I decide to debug what happen inside. However, it seems the calculation is happened on GPU, so it trigger segmentation fault, so I have to try to put variables on CPU.

However without any difference other than device placement, the following two piece of codes are outputting totally different:

```
def test5():
    from tensorflow.python import debug as tf_debug
    import numpy as np

    tf_labels = tf.placeholder(dtype=tf.int32, shape=[None])
    tf_logits = tf.placeholder(dtype=tf.int32, shape=[None])

    labels = np.array([1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], np.int32)
    logits = np.array([1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], np.int32)

    metrics = Metrics("accuracy")
    accuracy = metrics(tf_labels, tf_logits)

    writer = tf.summary.FileWriter("utils", tf.get_default_graph())
    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer(), tf.local_variables_initializer()])

        sess = tf_debug.LocalCLIDebugWrapperSession(sess)

        accu = sess.run(accuracy, feed_dict = {tf_labels: labels, tf_logits: logits})

        print(accu)

    writer.close()

def test6():
    from tensorflow.python import debug as tf_debug
    import numpy as np

    with tf.device('cpu:0'):
        labels = tf.constant([1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], tf.int32)
        logits = tf.constant([1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], tf.int32)

        metrics = Metrics("accuracy")
        accuracy = metrics(labels, logits)

    writer = tf.summary.FileWriter("utils", tf.get_default_graph())
    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer(), tf.local_variables_initializer()])

        # sess = tf_debug.LocalCLIDebugWrapperSession(sess)

        accu = sess.run(accuracy)

        print(accu)

    writer.close()

```

Without device placement directive, the output is correct, 0.88. With device placement directive, the output is wrong, 0.0. So dear deepminder, could you please tell me who is responsible for this issue, sonnet or tensorflow?

**Comment by malcolmreynolds** [comment]
@mingyr 

I think there are a few different things going on here - firstly, I can't reproduce your result about the device placement changing the result, but there are a number of things that could influence this. I can however reproduce the issue of getting a different result when using placeholders vs using tf.constant. The following code, I believe, does both the accuracy computations at once and I'm consistently seeing accuracy_ph = 0.0, accuracy_const=0.88:

```
def test7():
  labels_np = np.array([1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0,
                        0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]).astype(np.int32)
  logits_np = np.array([1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
                        0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]).astype(np.int32)

  labels_const = tf.constant(labels_np, tf.int32)
  logits_const = tf.constant(logits_np, tf.int32)
  labels_ph = tf.placeholder(dtype=tf.int32, shape=[None])
  logits_ph = tf.placeholder(dtype=tf.int32, shape=[None])

  accuracy_const, accuracy_update_const = tf.metrics.accuracy(
      labels_const, logits_const, name="acc_const")
  with tf.control_dependencies([accuracy_update_const]):
    accuracy_const = tf.identity(accuracy_const)

  accuracy_ph, accuracy_update_ph = tf.metrics.accuracy(
      labels_ph, logits_ph, name="acc_ph")
  with tf.control_dependencies([accuracy_update_ph]):
    accuracy_ph = tf.identity(accuracy_ph)

  with tf.Session() as session:
    session.run([tf.global_variables_initializer(),
                 tf.local_variables_initializer()])

    to_fetch = {
        "accuracy_const": accuracy_const,
        "accuracy_ph": accuracy_ph,
    }
    feed_dict = {
        labels_ph: labels_np,
        logits_ph: logits_np,
    }
    print(session.run(to_fetch, feed_dict))
```

Note that this doesnt use Sonnet, just raw TF  & tf.metrics, so I believe the problem is independent of Sonnet. I don't have time to dig further into this now but I suggest you open a ticket with the TF GitHub, referencing this one.
